src/util/logging.cpp

```diff
@@ -1,112 +1,161 @@
-            shouldFlush = Logging::flushing(LogLevel::Debug);
-            break;
-        case QtInfoMsg:
-            tag = "Info [";
-            baSize += strlen(tag);
-            shouldPrint = Logging::enabled(LogLevel::Info);
-            shouldFlush = Logging::flushing(LogLevel::Info);
-            break;
-        case QtWarningMsg:
-            tag = "Warning [";
-            baSize += strlen(tag);
-            shouldPrint = Logging::enabled(LogLevel::Warning);
-            shouldFlush = Logging::flushing(LogLevel::Warning);
-            break;
-        case QtCriticalMsg:
-            tag = "Critical [";
-            baSize += strlen(tag);
-            shouldFlush = true;
-            isDebugAssert = input.startsWith(QLatin1String(kDebugAssertPrefix));
-            break;
-        case QtFatalMsg:
-            tag = "Fatal [";
-            baSize += strlen(tag);
-            shouldFlush = true;
-            break;
-        default:
-            tag = "Unknown [";
-            baSize += strlen(tag);
+    case QtDebugMsg:
+        levelName = "Debug";
+        isControllerDebug =
+                input.startsWith(QLatin1String(
+                        ControllerDebug::kLogMessagePrefix));
+        if (isControllerDebug ||
+                Logging::enabled(LogLevel::Debug)) {
+            writeFlags |= WriteFlag::StdErr;
+            writeFlags |= WriteFlag::File;
+        }
+        if (Logging::shouldFlush(LogLevel::Debug)) {
+            writeFlags |= WriteFlag::Flush;
+        }
+        // TODO: Remove the following line.
+        // Do not write debug log messages into log file if log level
+        // Debug is not enabled starting with release 2.4.0! Until then
+        // write debug messages unconditionally into the log file
+        writeFlags |= WriteFlag::File;
+        break;
+    case QtInfoMsg:
+        levelName = "Info";
+        if (Logging::enabled(LogLevel::Info)) {
+            writeFlags |= WriteFlag::StdErr;
+        }
+        if (Logging::shouldFlush(LogLevel::Info)) {
+            writeFlags |= WriteFlag::Flush;
+        }
+        // Write unconditionally into log file
+        writeFlags |= WriteFlag::File;
+        break;
+    case QtWarningMsg:
+        levelName = "Warning";
+        if (Logging::enabled(LogLevel::Warning)) {
+            writeFlags |= WriteFlag::StdErr;
+        }
+        if (Logging::shouldFlush(LogLevel::Warning)) {
+            writeFlags |= WriteFlag::Flush;
+        }
+        // Write unconditionally into log file
+        writeFlags |= WriteFlag::File;
+        break;
+    case QtCriticalMsg:
+        levelName = "Critical";
+        writeFlags = WriteFlag::All;
+        isDebugAssert = input.startsWith(QLatin1String(kDebugAssertPrefix));
+        break;
+    case QtFatalMsg:
+        levelName = "Fatal";
+        writeFlags = WriteFlag::All;
+        break;
     }
+    VERIFY_OR_DEBUG_ASSERT(levelName) {
+        return;
+    }
+    int levelName_len = strlen(levelName);
```

Comment:
Not sure if the compiler can this optimize this away at this point. Maybe we can force it by a constexpr.

Reply:
I don't understand how and why this is a problem here? I will declare the local variable as const, but constexpr won't work. Counting <10 characters shouldn't cause any performance problems even if the optimizer doesn't recognize the constant strings.

Reply:
Reverted